// mimorelinks/libs/DeviceParser.js
const DEVICE_MAP = {
    'gateway': {name:'Gateway'},
    'gateway.v3':{name:'Gateway'},
    'magnet': {name:'Mi Window and Door Sensor', modelCode:'MCCGQ01LM'},
    'sensor_magnet':{name:'ContactSensor'},
    'motion': {name:'Motion Sensor', modelCode:'RTCGQ01LM'},
    'switch': {name:'Mi Wirelles Switch', modelCode:'WXKG01LM'},
    'sensor_ht': {name:'Mi Temperature And Humidity Sensor', modelCode:'WSDCGQ01LM'},
    'ctrl_neutral1': {name:'Aqara Wall Switch, Single Rocker', modelCode:'QBKG04LM'},
    'ctrl_neutral2': {name:'Aqara Wall Switch, Double Rocker', modelCode:'QBKG03LM'},
    'ctrl_ln1': {name:'Wall Switch, With Neutral, Single Rocker'},
    'ctrl_ln1.aq1': {name:'Aqara Wall Switch, With Neutral, Single Rocker', modelCode:'QBKG11LM'},
    'ctrl_ln2': {name:'Aqara Wall Switch, With Neutral, Double Rocker'},
    'ctrl_ln2.aq1': {name:'Aqara Wall Switch, With Neutral, Double Rocker', modelCode:'QBKG12LM'},
    '86sw1': {name:'Aqara Wirelles Single Switch', modelCode:'WXKG03LM'},
    '86sw2': {name:'Aqara Wirelles Double Switch', modelCode:'WXKG02LM'},
    'sensor_86sw1.aq1':{name:'SingleButton86W'},
    'sensor_86sw2.aq1':{name:'DuplexButton86W'},
    'plug': {name:'Mi Smart Plug', modelCode:'ZNCZ02LM'},
    '86plug': {name:'PlugBase86'},
    'ctrl_86plug':{name:'AqaraPlug86'},
    'ctrl_86plug.aq1':{name:'Aqara Wall Outlet', modelCode:'QBCZ11LM'},
    'cube': {name:'MagicSquare'},
    'sensor_cube':{name:'MagicSquareAqara'},
    'smoke': {name:'SmokeDetector'},
    'sensor_smoke':{name:'SmokeDetector'},
    'natgas': {name:'NatgasDetector'},
    'curtain': {name:'ElectricCurtain'},
    'sensor_magnet':{name:'ContactSensor', modelCode:'MCCGQ01LM'},
    'sensor_magnet.aq2': {name:'ContactSensor2'},
    'sensor_motion':{name:'MotionSensor', modelCode:'RTCGQ01LM'},
    'sensor_motion.aq2': {name:'MotionSensor2'},
    'sensor_switch':{name:'Button'},
    'sensor_switch.aq2': {name:'Button2'},
    'sensor_switch.aq3':{name:'Button3'},
    'weather':{name:'TemperatureAndHumiditySensor'},
    'weather.v1': {name:'TemperatureAndHumiditySensor2'},
    'sensor_wleak.aq1': {name:'WaterDetector', modelCode:'SJCGQ11LM'},
    'acpartner.v3':{name:'AqaraACPartner'},
    'sensor_natgas':{name:'NatgasDetector'}
};

const STATUS = {
    OPEN: 'open',
    CLOSE: 'close',
    MOTION: 'motion',
    CLICK: 'click',
    DOUBLE_CLICK: 'double_click',
    BOTH_CLICK: 'both_click',
    ON: 'on',
    OFF: 'off',
    ALARM:'alarm',
    NORMAL:'normal',
    UNKNOWN:'unknown',
    STOP:'stop',
    AUTO:'auto',
    FLIP90:'flip90',
    FLIP180:'flip180',
    MOVE:'move',
    TAP_TWICE:'tap_twice',
    SHAKE_AIR:'shake_air',
    SWING:'swing',
    ALERT:'alert',
    FREE_FALL:'free_fall',
    ROTATE:'rotate',
    SHAKE:'shake',
    LEAK:'leak'
};

let DeviceParser = {
    STATUS: STATUS,
    getNameByModel (model) {
        if (DEVICE_MAP.hasOwnProperty(model)) {
            return DEVICE_MAP[model];
        }
    }
};

module.exports = DeviceParser;
