// mimorelinks/libs/Device.js
class Device {
    constructor ({model, sid, short_id, data}) {
        this.model = model;
        this.sid = sid;
        this.short_id = short_id;
        this.data = data;
        this.voltage = data.voltage;
        this.channel_0 = data.channel_0;
        this.channel_1 = data.channel_1;
        this.dual_channel = data.dual_channel;
    }

    update (data) {
        for (let key in data) {
            if (this.hasOwnProperty(key)) {
                this[key] = data[key];
            }

            if (data.model == '86sw2') {
                this.channel_0 = false;
                this.channel_1 = false;
                this.dual_channel = false;
            }
        }

        for (let key in data.data) {
            this[key] = data.data[key];

            if(key == 'channel_0') {
                this.channel_0 = true;
            }

            if(key == 'channel_1') {
                this.channel_1 = true;
            }

            if(key == 'dual_channel') {
                this.dual_channel = true;
            }

        }
    }
}

module.exports = Device;
