// mimorelinks/libs/Device.js
class Device {
    constructor ({model, sid, short_id, data}) {
        this.model = model;
        this.sid = sid;
        this.short_id = short_id;
        this.data = data;
        this.voltage = data.voltage;
        this.channel_0 = data.channel_0;
        this.channel_1 = data.channel_1;
        this.dual_channel = data.dual_channel;
        this.status = data.status;
        this.no_motion = data.no_motion;
        this.leaked = false;
    }

    update (data) {
        for (let key in data) {
            if (this.hasOwnProperty(key)) {
                this[key] = data[key];
            }

            if (data.model == '86sw2') {
                this.channel_0 = false;
                this.channel_1 = false;
                this.dual_channel = false;
            }

            if (data.model == 'motion') {
                this.status = false;
            }

        }

        for (let key in data.data) {
            this[key] = data.data[key];

            if (key == 'channel_0') {
                this.channel_0 = true;
            }

            if (key == 'channel_1') {
                this.channel_1 = true;
            }

            if (key == 'dual_channel') {
                this.dual_channel = true;
            }

            if (key == 'status' && data.data[key] == 'motion') {
                this.status = true;
                this.no_motion = 0;
            }

            if (key == 'status' && data.data[key] == 'leak') {
                this.leaked = true;
            }

            if (key == 'status' && data.data[key] == 'no_leak') {
                this.leaked = false;            }
        }
    }
}

module.exports = Device;
