// mimorelinks/libs/GatewayHelper.js
const Gateway = require('./Gateway');
const utils = require('./utils');

class GatewayHelper {
    constructor (platform) {
        if (!platform) {
            throw new Error('[GatewayHelper:constructor] Param error');
        }
        this.gateways = {};
        this.platform = platform;
    }

    add (gateway) {
        if (!gateway || !gateway.sid || this.gateways.hasOwnProperty(gateway.sid)) {
            return;
        }
        this.gateways[gateway.sid] = gateway;
    }

    addOrUpdate (data) {
        let sid = data.sid;
        if (this.gateways.hasOwnProperty(sid)) {
            let gateway = this.gateways[sid];
            gateway.update(data);
        } else {
            this.gateways[sid] = new Gateway(data);
        }
    }

    remove (sid) {
        delete this.gateways[sid];
    }

    getBySid (sid) {
        return this.gateways[sid];
    }

    uploadBySid (sid, data) {
        if (this.gateways.hasOwnProperty(sid)) {
            let gateway = this.gateways[sid];
            gateway.update(data);
        }
    }

    getGatewayList () {
        let list = [];
        for (let key in this.gateways) {
            let value = this.gateways[key];
            list.push(value);
        }
        return list;
    }

    getAll () {
        return this.gateways;
    }

    getIdList (sid) {
        let gateway = this.getBySid(sid);
        if (gateway) {
            this.platform.send(gateway.ip, gateway.port, {
                cmd: 'get_id_list'
            });
        } else {
            console.error('[GatewayHelper:getIdList] sid:%s is not exit', sid);
        }
    }

    read (sid) {
        console.log('[GatewayHelper:read] sid=%s', sid);
        let gateway = this.getBySid(sid);
        if (gateway) {
            this.platform.send(gateway.ip, gateway.port, {
                cmd: 'read',
                sid: sid
            });
        } else {
            console.error('[GatewayHelper:read] sid:%s can not find gateway', sid);
        }
    }

    write (sid, data) {
        console.log('[GatewayHelper:write] sid=%s', sid);
        let gateway = this.getBySid(sid);
        if (gateway) {
            let msg = {
                cmd: 'write',
                model: 'gateway',
                sid: gateway.sid,
                data: Object.assign({}, data)
            };
            msg.data.key = utils.cipher(gateway.token, gateway.password, gateway.iv);
            this.platform.send(gateway.ip, gateway.port, msg);
        } else {
            console.error('[GatewayHelper:read] sid:%s can not find', sid);
        }
    }

    controlLightPower({sid, power}) {
        let prepValue = 0;
        let hue=0, saturation=0, brightness=0
        if(power) {
            if(!hue) {
                hue = 0;
            }
            if(!saturation) {
                saturation = 0 * 100;
            }
            if(!brightness) {
                brightness = 50;
            }
            let rgb = utils.hsb2rgb([hue, saturation/100, 1]);
            prepValue = parseInt(utils.dec2hex(brightness, 2) + utils.dec2hex(rgb[0], 2) + utils.dec2hex(rgb[1], 2) + utils.dec2hex(rgb[2], 2), 16);
        }
        console.log('prepValue:',prepValue);
        this.write(sid, {rgb: prepValue});
    }

    controlLightHLS({sid, hue, saturation, brightness}) {
        let prepValue = 0;

        if(!hue) {
            hue = 0;
        }

        if(!saturation) {
            saturation = 0 * 100;
        }
        if(!brightness) {
            brightness = 50;
        }

        if(hue > 100){
            hue = 100;
        }

        if(saturation > 100){
            saturation = 100;
        }

        if(brightness > 100){
            brightness = 100;
        }

        let rgb = utils.hsb2rgb([hue, saturation/100, 1]);
        prepValue = parseInt(utils.dec2hex(brightness, 2) + utils.dec2hex(rgb[0], 2) + utils.dec2hex(rgb[1], 2) + utils.dec2hex(rgb[2], 2), 16);
        console.log('prepValue:',prepValue);
        this.write(sid, {rgb: prepValue});
    }

    controlLightRGB({sid, rgb}) {
        let prepValue = 0;
        if(rgb) {
            prepValue = rgb;
        }
        this.write(sid, {rgb: prepValue});
    }

    controlMid({sid, mid}) {
        let prepValue = 1000;
        if(mid) {
            prepValue = mid;
        }
        this.write(sid, {mid: prepValue});
    }
}

module.exports = GatewayHelper;
